src/LibraryManager/Manifest.cs

```diff
@@ -127,6 +127,210 @@ private static bool IsValidManifestVersion(string version)
+            }
+
+            IProvider provider = null;
+            if (string.IsNullOrEmpty(providerId) && string.IsNullOrEmpty(DefaultProvider))
+            {
+                errors.Add(PredefinedErrors.ProviderIsUndefined());
+            }
+            else
+            {
+                providerId = string.IsNullOrEmpty(providerId) ? DefaultProvider : providerId;
+
+                provider = _dependencies.Providers.FirstOrDefault(p => p.Id == providerId);
+                if (provider == null)
+                {
+                    errors.Add(PredefinedErrors.ProviderUnknown(providerId));
+                }
+            }
+
+            if (errors.Any())
+            {
+                result = new LibraryInstallationResult(errors);
+                return result;
+            }
+
+            var desiredState = new LibraryInstallationState()
+            {
+                LibraryId = libraryId,
+                Files = files,
+                ProviderId = providerId,
+                DestinationPath = destination
+            };
+
+            if (CheckAlreadyInstalled(desiredState))
+            {
+                errors.Add(PredefinedErrors.LibraryAlreadyInstalled(libraryId, providerId));
+            }
+
+            if (errors.Any())
+            {
+                result = new LibraryInstallationResult(errors);
+                return result;
+            }
+
+            result = await provider.InstallAsync(desiredState, cancellationToken);
+
+            // Remove destination and provider if they match the defaults;
+            if (desiredState.DestinationPath == DefaultDestination)
+            {
+                desiredState.DestinationPath = null;
+            }
+
+            if (desiredState.ProviderId == DefaultProvider)
+            {
+                desiredState.ProviderId = null;
+            }
+
+            if (result.Success)
+            {
+                _libraries.Add(desiredState);
+            }
+
+            return result;
+        }
+
+        private bool CheckAlreadyInstalled(LibraryInstallationState desiredState)
+        {
+            // compare id, provider and destination
+            return Libraries.Any(l => l.LibraryId == desiredState.LibraryId
+                        && (l.DestinationPath == desiredState.DestinationPath)
+                        && (l.ProviderId == desiredState.ProviderId 
+                            || (l.ProviderId == null && desiredState.ProviderId == DefaultProvider)));
+        }
+
+        /// <summary>
+        /// 
+        /// </summary>
+        /// <param name="libraryToUpdate"></param>
+        /// <param name="newId"></param>
+        /// <param name="deleteFileAction"></param>
+        /// <param name="cancellationToken"></param>
+        /// <returns></returns>
+        public async Task<ILibraryInstallationResult> UpdateLibraryAsync(
+            ILibraryInstallationState libraryToUpdate,
```

Comment:
Update should just figure out the new version is and call uninstall(ol version)  InstallLibrary(new version), no?

Reply:
No, we need to first make sure that the new Version can be installed. For cases when new version won't install (eg. the files specified in the old version are no longer valid for the new version), we shouldn't delete the files on disk.

Reply:
There is a bigger work item for us that (applies to all operations): we need to come up to a plan to "recover from failure". This would restore the files to prvious status in case any operation fails. We can talk about this early next week. 
To determine the delta between "before" and "after" files, take look at the method 

private async Task RemoveFilesAsync(Manifest newManifest)
https://github.com/aspnet/LibraryManager/blob/master/src/LibraryManager.Vsix/Json/TextviewCreationListener.cs
 
We can move it outside of TextviewCreationListener so we can all have the same implementation